File: frontend/src/lib/goldScoring.ts
/**
 * Moteur de scoring GOLD (PUR, miroir EXACT du backend `claire/projects/gold_scoring.py`).
 *
 * Décide, PAR PHRASE, le gold à partir des votes : électorat PONDÉRÉ où les annotateurs
 * pèsent PLUS que les LLM. La parité TS/PY est verrouillée par un golden partagé
 * (cf. dossier docs/pactiva/dossier-gold-tech/02-scoring-engine.md, ADR-003).
 *
 * - Les ANNOTATEURS font autorité ; les LLM aident selon `llmRole`.
 * - Décision humaine ≠ consensus LLM (`humanDissent`) = SIGNAL FORT → jamais d'auto.
 * - Classification strict / majority / divergence (sur les annotateurs).
 * - Niveaux d'auto : auto_1click (accord absolu) / auto (peu risqué) / manual.
 */

export const ANNOTATOR_WEIGHT_DEFAULT = 3.0;
export const LLM_WEIGHT_DEFAULT = 1.0;
export const LOW_CONFIDENCE = 0.34;
export const SECONDARY_MIN_ANNOTATORS = 2;
export const LLM_UNANIMOUS_MIN = 3;
export const MAJORITY_RATIO = 2 / 3;

export type LlmRole = "ignore" | "tiebreak" | "signal" | "full";
export type AgreementClass = "strict" | "majority" | "divergence" | "empty";
export type RiskBand = "low" | "medium" | "high";
export type AutoLevel = "auto_1click" | "auto" | "manual";

export interface Vote {
  voterId: string;
  primary: string | null; // null = phrase non couverte par cette source
  secondaries?: string[];
  isLlm?: boolean;
}

export interface ScoringConfig {
  annotatorWeight: number;
  llmWeight: number;
  llmRole: LlmRole;
  perAnnotator: Record<string, number>;
  perLlm: Record<string, number>;
  secondaryMinAnnotators: number;
  reliability: Record<string, number>;
}

export interface GoldScore {
  primary: string | null;
  secondaries: string[];
  agreementClass: AgreementClass;
  confidence: number;
  riskBand: RiskBand;
  humanDissent: boolean;
  humanBlock: string | null;
  llmBlock: string | null;
  autoLevel: AutoLevel;
  tally: Record<string, number>;
}

export function defaultConfig(): ScoringConfig {
  return {
    annotatorWeight: ANNOTATOR_WEIGHT_DEFAULT,
    llmWeight: LLM_WEIGHT_DEFAULT,
    llmRole: "tiebreak",
    perAnnotator: {},
    perLlm: {},
    secondaryMinAnnotators: SECONDARY_MIN_ANNOTATORS,
    reliability: {},
  };
}

const clamp01 = (x: number) => (x < 0 ? 0 : x > 1 ? 1 : x);

function weightOf(v: Vote, cfg: ScoringConfig): number {
  if (v.isLlm) return cfg.perLlm[v.voterId] ?? cfg.llmWeight;
  return cfg.perAnnotator[v.voterId] ?? cfg.annotatorWeight;
}

function tally(votes: Vote[], cfg: ScoringConfig): Record<string, number> {
  const out: Record<string, number> = {};
  for (const v of votes) {
    if (v.primary === null || v.primary === undefined) continue;
    out[v.primary] = (out[v.primary] ?? 0) + weightOf(v, cfg);
  }
  return out;
}

/** Argmax avec départage LEXICOGRAPHIQUE (parité exacte avec le backend). */
function argmax(t: Record<string, number>): string | null {
  let best: string | null = null;
  let bestW = -Infinity;
  for (const k of Object.keys(t).sort()) {
    const w = t[k] ?? 0;
    if (w > bestW) {
      bestW = w;
      best = k;
    }
  }
  return best;
}

const sum = (t: Record<string, number>) => Object.values(t).reduce((a, b) => a + b, 0);

export function scoreSentence(votes: Vote[], config?: Partial<ScoringConfig>): GoldScore {
  const cfg: ScoringConfig = { ...defaultConfig(), ...(config ?? {}) };
  const validRoles: LlmRole[] = ["ignore", "tiebreak", "signal", "full"];
  const role: LlmRole = validRoles.includes(cfg.llmRole) ? cfg.llmRole : "tiebreak";

  const norm = (v: Vote): Vote => ({ ...v, secondaries: v.secondaries ?? [] });
  const all = votes.map(norm);
  const humans = all.filter((v) => !v.isLlm);
  const llms = all.filter((v) => v.isLlm);
  const coveringHumans = humans.filter((v) => v.primary !== null);
  const coveringLlms = llms.filter((v) => v.primary !== null);

  const humanTally = tally(humans, cfg);
  const llmTally = tally(llms, cfg);
  const humanBlock = argmax(humanTally);
  const llmBlock = argmax(llmTally);

  // ── Électorat de décision selon le rôle des LLM ──
  let decision: Record<string, number>;
  if (role === "full") {
    decision = { ...humanTally };
    for (const [k, w] of Object.entries(llmTally)) decision[k] = (decision[k] ?? 0) + w;
  } else if (role === "tiebreak") {
    decision = { ...humanTally };
    const vals = Object.values(humanTally);
    if (vals.length > 0) {
      const top = Math.max(...vals);
      const tied = Object.keys(humanTally).filter((k) => humanTally[k] === top);
      if (tied.length > 1) for (const k of tied) decision[k] = (decision[k] ?? 0) + (llmTally[k] ?? 0);
    }
  } else {
    decision = { ...humanTally }; // ignore | signal
  }
  if (Object.keys(decision).length === 0 && role !== "ignore" && Object.keys(llmTally).length > 0) {
    decision = { ...llmTally };
  }

  const primary = argmax(decision);

  // ── Confiance = soutien pondéré du primaire sur l'électorat de confiance ──
  // Humains toujours ; LLM sauf rôle 'ignore' (un renfort LLM concordant remonte la confiance).
  const supportVotes = coveringHumans.concat(role === "ignore" ? [] : coveringLlms);
  const supportTotal = supportVotes.reduce((acc, v) => acc + weightOf(v, cfg), 0);
  const supportTop = supportVotes
    .filter((v) => v.primary === primary)
    .reduce((acc, v) => acc + weightOf(v, cfg), 0);
  const support = supportTotal > 0 ? supportTop / supportTotal : 0;
  const reliability = primary ? cfg.reliability[primary] ?? 1.0 : 0;
  const confidence = Math.round(clamp01(support) * reliability * 10000) / 10000;

  // ── Signal fort ──
  const humanDissent = humanBlock !== null && llmBlock !== null && humanBlock !== llmBlock;

  // ── Classification (sur les annotateurs) ──
  let agreementClass: AgreementClass;
  const secKey = (v: Vote) => [...new Set(v.secondaries ?? [])].sort().join("");
  if (coveringHumans.length === 0) {
    agreementClass = "empty";
  } else {
    const primaries = new Set(coveringHumans.map((v) => v.primary));
    const secSets = new Set(coveringHumans.map(secKey));
    if (primaries.size === 1 && secSets.size === 1) {
      agreementClass = "strict";
    } else {
      const humanTotal = sum(humanTally);
      if (humanBlock !== null && (humanTally[humanBlock] ?? 0) > humanTotal / 2) agreementClass = "majority";
      else agreementClass = "divergence";
    }
  }

  // ── Secondaires (≥ N annotateurs, hors primaire) ──
  const secCounts: Record<string, number> = {};
  for (const v of coveringHumans) {
    for (const s of new Set(v.secondaries ?? [])) if (s !== primary) secCounts[s] = (secCounts[s] ?? 0) + 1;
  }
  const secondaries = Object.keys(secCounts)
    .filter((s) => (secCounts[s] ?? 0) >= cfg.secondaryMinAnnotators)
    .sort();

  // ── Bande de risque ──
  let riskBand: RiskBand;
  if (agreementClass === "divergence" || humanDissent || confidence < LOW_CONFIDENCE) riskBand = "high";
  else if (agreementClass === "strict" && !humanDissent) riskBand = "low";
  else riskBand = "medium";

  // ── Niveau d'auto-résolution ──
  const secondariesIdentical =
    coveringHumans.length > 0 && new Set(coveringHumans.map(secKey)).size === 1;
  const llmUnanimous =
    coveringLlms.length >= LLM_UNANIMOUS_MIN && new Set(coveringLlms.map((v) => v.primary)).size === 1;
  const ratio = coveringHumans.length
    ? coveringHumans.filter((v) => v.primary === humanBlock).length / coveringHumans.length
    : 0;
  let autoLevel: AutoLevel;
  if (humanDissent) autoLevel = "manual";
  else if (agreementClass === "strict" && secondariesIdentical) autoLevel = "auto_1click";
  else if (ratio >= MAJORITY_RATIO - 1e-9 && llmUnanimous && llmBlock === humanBlock) autoLevel = "auto";
  else autoLevel = "manual";

  return {
    primary,
    secondaries,
    agreementClass,
    confidence,
    riskBand,
    humanDissent,
    humanBlock,
    llmBlock,
    autoLevel,
    tally: decision,
  };
}

File: backend/claire/projects/gold_scoring.py
"""Moteur de scoring GOLD (PUR) — résolution de conflits inter-annotateurs.

À côté d'iaa.py / concordance.py : aucune dépendance Django, testable en unité + property-based.
Décide, PAR PHRASE, le gold à partir des votes : électorat PONDÉRÉ où les annotateurs pèsent
PLUS que les LLM (dossier docs/pactiva/dossier-gold-tech/02-scoring-engine.md, ADR-003).

Principes :
- Les ANNOTATEURS font autorité (le gold est humain) ; les LLM aident selon `llm_role`.
- Une décision humaine ≠ consensus LLM (`human_dissent`) est un SIGNAL FORT → jamais d'auto.
- Classification : strict / majority / divergence (sur les annotateurs).
- Niveaux d'auto-résolution : auto_1click (accord absolu) / auto (cas peu risqué) / manual.
- Multi-label : primaire + secondaires décidés séparément.
"""

from __future__ import annotations

from dataclasses import dataclass

ANNOTATOR_WEIGHT_DEFAULT = 3.0
LLM_WEIGHT_DEFAULT = 1.0
LOW_CONFIDENCE = 0.34
SECONDARY_MIN_ANNOTATORS = 2
LLM_UNANIMOUS_MIN = 3
MAJORITY_RATIO = 2.0 / 3.0

LLM_ROLES = ("ignore", "tiebreak", "signal", "full")


@dataclass(frozen=True)
class Vote:
    """Vote d'une source pour UNE phrase. `primary=None` = phrase non couverte par cette source."""
    voter_id: str
    primary: str | None
    secondaries: tuple[str, ...] = ()
    is_llm: bool = False


@dataclass
class GoldScore:
    primary: str | None
    secondaries: list[str]
    agreement_class: str  # strict | majority | divergence | empty
    confidence: float
    risk_band: str  # low | medium | high
    human_dissent: bool
    human_block: str | None
    llm_block: str | None
    auto_level: str  # auto_1click | auto | manual
    tally: dict  # thème -> masse de poids (électorat de décision)


def default_config() -> dict:
    """Config par défaut « confiance aux annotateurs » (cf. presets.yaml)."""
    return {
        "annotator_weight": ANNOTATOR_WEIGHT_DEFAULT,
        "llm_weight": LLM_WEIGHT_DEFAULT,
        "llm_role": "tiebreak",
        "per_annotator": {},   # voter_id -> poids
        "per_llm": {},         # voter_id -> poids
        "secondary_min_annotators": SECONDARY_MIN_ANNOTATORS,
        "reliability": {},     # thème -> kappa de fiabilité (∈[0,1]) ; défaut 1.0
    }


def _clamp01(x: float) -> float:
    return 0.0 if x < 0 else 1.0 if x > 1 else x


def _weight(vote: Vote, config: dict) -> float:
    if vote.is_llm:
        return float(config.get("per_llm", {}).get(vote.voter_id, config.get("llm_weight", LLM_WEIGHT_DEFAULT)))
    return float(config.get("per_annotator", {}).get(vote.voter_id, config.get("annotator_weight", ANNOTATOR_WEIGHT_DEFAULT)))


def _tally(votes: list[Vote], config: dict) -> dict:
    """Masse de poids par thème PRIMAIRE (ignore les votes non couvrants)."""
    out: dict[str, float] = {}
    for v in votes:
        if v.primary is None:
            continue
        out[v.primary] = out.get(v.primary, 0.0) + _weight(v, config)
    return out


def _argmax(tally: dict) -> str | None:
    """Thème de masse maximale ; départage LEXICOGRAPHIQUE (déterminisme/idempotence)."""
    if not tally:
        return None
    return max(sorted(tally), key=lambda k: tally[k])


def _block(votes: list[Vote], config: dict) -> tuple[str | None, dict]:
    t = _tally(votes, config)
    return _argmax(t), t


def score_sentence(votes: list[Vote], config: dict | None = None) -> GoldScore:
    """Calcule la décision gold d'une phrase à partir des votes (annotateurs + LLM). PUR."""
    cfg = {**default_config(), **(config or {})}
    role = cfg.get("llm_role", "tiebreak")
    if role not in LLM_ROLES:
        role = "tiebreak"

    humans = [v for v in votes if not v.is_llm]
    llms = [v for v in votes if v.is_llm]
    covering_humans = [v for v in humans if v.primary is not None]
    covering_llms = [v for v in llms if v.primary is not None]

    human_block, human_tally = _block(humans, cfg)
    llm_block, llm_tally = _block(llms, cfg)

    # ── Électorat de DÉCISION selon le rôle des LLM ──────────────────────────
    if role == "full":
        decision = dict(human_tally)
        for k, w in llm_tally.items():
            decision[k] = decision.get(k, 0.0) + w
    elif role == "tiebreak":
        decision = dict(human_tally)
        # LLM ne départagent QUE les ex-aequo humains de tête.
        if human_tally:
            top = max(human_tally.values())
            tied = [k for k, w in human_tally.items() if w == top]
            if len(tied) > 1:
                for k in tied:
                    decision[k] = decision.get(k, 0.0) + llm_tally.get(k, 0.0)
    else:  # ignore | signal : les LLM ne participent pas à la décision
        decision = dict(human_tally)

    # Repli LLM si AUCUN humain n'a couvert la phrase (rare) et rôle non-ignore.
    if not decision and role in ("full", "tiebreak", "signal") and llm_tally:
        decision = dict(llm_tally)

    primary = _argmax(decision)

    # ── Confiance = soutien pondéré du primaire sur l'électorat de confiance ──
    # Humains toujours ; LLM SAUF rôle 'ignore' (un renfort LLM concordant remonte la
    # confiance et abaisse le risque, cohérent avec l'auto-résolution des cas peu risqués).
    support_votes = covering_humans + ([] if role == "ignore" else covering_llms)
    support_total = sum(_weight(v, cfg) for v in support_votes)
    support_top = sum(_weight(v, cfg) for v in support_votes if v.primary == primary)
    support = (support_top / support_total) if support_total > 0 else 0.0
    reliability = float(cfg.get("reliability", {}).get(primary, 1.0)) if primary else 0.0
    confidence = _clamp01(support) * reliability

    # ── Signal fort : décision humaine ≠ consensus LLM ───────────────────────
    human_dissent = (
        human_block is not None and llm_block is not None and human_block != llm_block
    )

    # ── Classification (sur les ANNOTATEURS, autorité du gold) ───────────────
    if not covering_humans:
        agreement_class = "empty"
    else:
        primaries = {v.primary for v in covering_humans}
        sec_sets = {tuple(sorted(set(v.secondaries))) for v in covering_humans}
        if len(primaries) == 1 and len(sec_sets) == 1:
            agreement_class = "strict"
        else:
            human_total = sum(human_tally.values())
            if human_block is not None and human_tally.get(human_block, 0.0) > human_total / 2.0:
                agreement_class = "majority"
            else:
                agreement_class = "divergence"

    # ── Secondaires : portés par ≥ N annotateurs (hors primaire) ─────────────
    sec_min = int(cfg.get("secondary_min_annotators", SECONDARY_MIN_ANNOTATORS))
    sec_counts: dict[str, int] = {}
    for v in covering_humans:
        for s in set(v.secondaries):
            if s != primary:
                sec_counts[s] = sec_counts.get(s, 0) + 1
    secondaries = sorted([s for s, n in sec_counts.items() if n >= sec_min])

    # ── Bande de risque ──────────────────────────────────────────────────────
    if agreement_class == "divergence" or human_dissent or confidence < LOW_CONFIDENCE:
        risk_band = "high"
    elif agreement_class == "strict" and not human_dissent:
        risk_band = "low"
    else:
        risk_band = "medium"

    # ── Niveau d'auto-résolution (politique auto-resolution.yaml) ────────────
    secondaries_identical = (
        len({tuple(sorted(set(v.secondaries))) for v in covering_humans}) == 1
        if covering_humans else False
    )
    llm_unanimous = (
        len(covering_llms) >= LLM_UNANIMOUS_MIN
        and len({v.primary for v in covering_llms}) == 1
    )
    # Ratio sur le NOMBRE d'annotateurs (fidèle à « 2/3 des annotateurs d'accord »).
    ratio = (
        sum(1 for v in covering_humans if v.primary == human_block) / len(covering_humans)
        if covering_humans else 0.0
    )
    if human_dissent:
        auto_level = "manual"
    elif agreement_class == "strict" and secondaries_identical:
        auto_level = "auto_1click"
    elif ratio >= MAJORITY_RATIO - 1e-9 and llm_unanimous and llm_block == human_block:
        auto_level = "auto"
    else:
        auto_level = "manual"

    return GoldScore(
        primary=primary,
        secondaries=secondaries,
        agreement_class=agreement_class,
        confidence=round(confidence, 4),
        risk_band=risk_band,
        human_dissent=human_dissent,
        human_block=human_block,
        llm_block=llm_block,
        auto_level=auto_level,
        tally=decision,
    )
